packages/flutter/lib/src/material/button_theme.dart

```diff
@@ -284,6 +330,382 @@ class ButtonThemeData extends Diagnosticable {
+      case ButtonTextTheme.primary: {
+        final Color fillColor = getFillColor(button);
+        final bool fillIsDark = fillColor != null
+          ? ThemeData.estimateBrightnessForColor(fillColor) == Brightness.dark
+          : getBrightness(button) == Brightness.dark;
+        if (fillIsDark)
+          return Colors.white;
+        if (button is FlatButton || button is OutlineButton)
+          return colorScheme.primary;
+        return Colors.black;
+      }
+    }
+
+    assert(false);
+    return null;
+  }
+
+  /// The color of the ink "splash" overlay that appears when the (enabled)
+  /// [button] is tapped.
+  ///
+  /// Returns the button's [MaterialButton.splashColor] if it is non-null.
+  ///
+  /// Otherwise, returns [splashColor] if [button] is a [RaisedButton] or an
+  /// [OutlineButton] and [splashColor] is non-null.
+  ///
+  /// Otherwise, returns [splashColor] if [button] is a [FlatButton],
+  /// [getTextTheme] is not [ButtonTextTheme.primary], and [splashColor] is
+  /// non-null.
+  ///
+  /// Otherwise, returns [getTextColor] with an opacity of 0.12.
+  Color getSplashColor(MaterialButton button) {
+    if (button.splashColor != null)
+      return button.splashColor;
+
+    if (splashColor != null && (button is RaisedButton || button is OutlineButton))
+      return splashColor;
+
+    if (splashColor != null && button is FlatButton) {
+      switch (getTextTheme(button)) {
+        case ButtonTextTheme.normal:
+        case ButtonTextTheme.accent:
+          return splashColor;
+        case ButtonTextTheme.primary:
+          break;
+      }
+    }
+
+    return getTextColor(button).withOpacity(0.12);
+  }
+
+  /// The color of the overlay that appears when the [button] is pressed.
+  ///
+  /// Returns the button's [MaterialButton.highlightColor] if it is non-null.
+  /// Otherwise the highlight color depends on [getTextTheme]:
+  ///
+  ///  * [ButtonTextTheme.normal], [ButtonTextTheme.accent]: returns
+  ///    [highlightColor] if it is non-null, otherwise the value
+  ///    of [getTextColor] with opacity 0.16.
+  ///  * [ButtonTextTheme.primary], returns [Colors.transparent].
+  Color getHighlightColor(MaterialButton button) {
+    if (button.highlightColor != null)
+      return button.highlightColor;
+
+    switch (getTextTheme(button)) {
+      case ButtonTextTheme.normal:
+      case ButtonTextTheme.accent:
+        return highlightColor ?? getTextColor(button).withOpacity(0.16);
+      case ButtonTextTheme.primary:
+        return Colors.transparent;
+    }
+
+    assert(false);
+    return Colors.transparent;
+  }
+
+  /// The [button]'s elevation when it is enabled and has not been pressed.
+  ///
+  /// Returns the button's [MaterialButton.elevation] if it is non-null.
+  ///
+  /// If button is a [FlatButton] then elevation is 0.0, otherwise it is 2.0.
+  double getElevation(MaterialButton button) {
+    if (button.elevation != null)
+      return button.elevation;
+    if (button is FlatButton)
```

Comment:
As long as button is exactly a FlatButton, I think it is faster to check the runtimeType than use an `is` check.  Since the latter has to verify the full type hierarchy.

You could also pass the type as method generic:

```
double getElevation<T extends materialButton>(T button) {
  if (T == FlatButton) { ... }
}
```

Reply:
I guess the above wouldn't work if you wanted to delegate to MaterialButton, and it might not be obvious but inference could mess up users here:

class MyButton extends MaterialButton {

   Widget build(_) => RawMaterialButton(foo: theme.getFoo(this)) // <- T is inferred as MyButton instead of MaterialButton
}

Reply:
If someone wants to write a button that's configured like a FlatButton but builds a little differently, they could subclass FlatButton and override its build method. FlatButtonWithIcon is essentially that.

The generic approach would be a little cheaper, and it would obviate creating classes like FlatButtonWithIcon. It's similar to an earlier approach I'd taken, where a value that represented the button's type was passed along as a MaterialButton property. It makes using the API a bit clunkier, although it also clarifies the meaning of each getFoo(button) method: getElevation<FlatButton>(myFlatButton) is a little clearer that we want the FlatButton elevation.

Reply:
I think the advantage of the is checks is that they will just work.  If it's slow we can always revisit